src/raft.rs

```diff
@@ -2215,14 +2355,103 @@ impl<T: Storage> Raft<T> {
         self.lead_transferee = None;
     }
 
+    // Send snapshot request message to the leader
     fn send_request_snapshot(&mut self) {
+        let m = self.prepare_request_snapshot(self.leader_id);
+        self.send(m);
+    }
+
+    // Prepare a Message for requesting snapshot from given `to`
+    fn prepare_request_snapshot(&mut self, to: u64) -> Message {
         let mut m = Message::default();
         m.set_msg_type(MessageType::MsgAppendResponse);
         m.index = self.raft_log.committed;
         m.reject = true;
         m.reject_hint = self.raft_log.last_index();
-        m.to = self.leader_id;
+        m.to = to;
         m.request_snapshot = self.pending_request_snapshot;
-        self.send(m);
+        m
+    }
+
+    #[inline]
+    fn pr_become_snapshot_and_log(
+        &self,
+        id: u64,
+        progress: &mut Progress,
+        snapshot_index: u64,
+        snapshot_term: u64,
+    ) {
+        debug!(
+            self.logger,
+            "[firstindex: {first_index}, commit: {committed}] sent snapshot[index: {snapshot_index}, term: {snapshot_term}] to {to}",
+            first_index = self.raft_log.first_index(),
+            committed = self.raft_log.committed,
+            snapshot_index = snapshot_index,
+            snapshot_term = snapshot_term,
+            to = id;
+            "progress" => ?progress,
+        );
+        progress.become_snapshot(snapshot_index);
+        debug!(
+            self.logger,
+            "paused sending replication messages to {}",
+            id;
+            "progress" => ?progress,
+        );
+    }
+
+    // Pick a delegate of the given group
+    //
+    // The delegate must satisfy conditions below:
+    // 1. Must be 'recent_active'
+    // 2. The progress state should be 'Replicate' but not 'paused'
+    // 3. The progress has smallest 'match'
+    // 4. Not requiring a snapshot
+    //
+    // Especially if all the members in a group needs snapshot, choose the delegate randomly.
+    //
+    // Why using the peer with smallest 'match' as a delegate ?
+    //
+    // Our target is to reduce the workload in the leader which means the group
+    // delegate should be able to send entries as much as possible. Therefore,
+    // the delegate must not be requiring a snapshot (either or requesting
+    // a snapshot) from the perspective of leader. And the more entries the delegate
+    // is able to send to the other group members, the lower pressure the leader has
+    // syncing raft logs to peers in typical log replication.
+    // Also, A peer with smaller 'match' is able to receive more un-compacted entries from leader
+    // and then send them to others in same group.
+    //
+    fn pick_delegate(&self, group: &[u64], prs: &ProgressSet) -> u64 {
```

Comment:
Seems in the new implementation, choose the peer with most raft logs is better?

Reply:
It seems more reasonable. Actually maybe we can choose any unpaused peer?